### ISSUE TITLE
Documentation: Custom matches functions

### ISSUE THREAD

**Issue by nickyblissAviva** [dequelabs/axe-core#2271]
Expectation: Clear documentation on where the path for the custom matches function is relative too

Actual: Current documentation just states relative path

Motivation: We are currently writing some custom matches function to filter out some known false positives that axe is unable to cope with but retain the testing for all other elements rather than turning the test off completely. We have tried a number of different possible relative paths but none seem to be changing the results returned so must be being ignored.

<pre><code>
axe-core version: 3.5.4

Browser and Assistive Technology versions

For Tooling issues:
- Node version: 12.13.1
- Platform:  Windows
</code></pre>

**Comment by straker** [comment]
Thanks for the issue. I'll see what it would take to make the docs more clear.

**Comment by nickyblissAviva** [comment]
@straker Would you be able to advise where the path should be relative from. I have tried a number of different possibilities but can't get any to work it just keeps using the original matching function

**Comment by nickyblissAviva** [comment]
We are currently calling axe core from cypress which is using:
```
axe.run( document, {
	runOnly : {
		type : 'tags',
		values : [ 'wcag2a', 'wcag2aa', 'best-practice' ], // Prob want to include Best Practice here once it doesn't blow up
	},
	rules : {
		region : { enabled : false }, // Returns potentially false positives if regions are not top level elements
		'color-contrast' : {
			matches : 'axe-rules/color-contrast-matches.js', // Path relative to where the run call is being made
		},
});
```
Custom matches function is ignored and default run

**Comment by straker** [comment]
So I dug into the code and `matches` does not accept a relative path for `axe.run`.  The only place I found a reference to `matches` being a relative path was in the [developer docs](https://github.com/dequelabs/axe-core/blob/develop/doc/developer-guide.md#rules), which is used for working on axe core itself (and also out of date now).

Instead, `matches` should be the function itself.

**Comment by nickyblissAviva** [comment]
@straker so I could still use matches but it must be a function not a string?
```
'color-contrast' : {
	matches : function( args ) {
		....
	},
},
```

**Comment by straker** [comment]
Correct

**Comment by nickyblissAviva** [comment]
That doesn't seem to work. Even just returning true of false within the function and doing nothing else the results I am getting back are the same :(

**Comment by straker** [comment]
That should work. What does your function look like? Also, is there a page I could test it on?

**Comment by nickyblissAviva** [comment]
I have just used this as the options passed into run as a very quick test:
```
 {
	runOnly : {
		type : 'tags',
		values : [ 'wcag2a', 'wcag2aa', 'best-practice' ],
	},
	rules : {
	'color-contrast' : {
		matches : function( args ) {
			return false;
		},
	},
},
```

This page you should be able to get too. The selected checkbox should cause a colour contrast violation even thought the white text is on an accessible green (ignore the faded out disabled state as it also doesn't like that but we dont seem to have another radio or checkbox in a selected state to demonstrate the issue.
https://standards.aviva.com/framework/element-library/v4/forms/checkbox/

**Comment by straker** [comment]
Ahh. I see what the problem is. Sorry I didn't notice it before. So to configure a rule to change the `matches` (or any other property), you have to use `axe.configure` to pass the `rules` property and not `axe.run`. The docs [say you can do this in `axe.run`](https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#options-parameter), but this is incorrect. You can only configure the `enabled` property in the `axe.run` options.

```js
axe.configure({
  rules: [{
    id: 'color-contrast',
    matches: function() { /* ... */ }
  }]
});

axe.run({
  runOnly : {
    type : 'tags',
    values : [ 'wcag2a', 'wcag2aa', 'best-practice' ],
  }
});
```

I'll create a ticket to go over all the docs before our 4.0 release coming up and make sure they are all in order.

**Comment by nickyblissAviva** [comment]
Thank you. I'll give that a go and see how we get on :)

**Comment by nickyblissAviva** [comment]
@straker is there a way of passing back to the default matches function within the custom one. Say we do our matches to exclude the few bits we want to exclude and then pass it back to the default function to complete the standard checks?

**Comment by straker** [comment]
Unfortunately we don't have a way to do that

**Comment by nickyblissAviva** [comment]
Hmm. Thats going to make it interesting to do then as I don't really want to loose all the matching functionality already there just exclude a few scenarios from that one rule.
Any suggestions?

**Comment by straker** [comment]
Ya, that matches has a lot going on. I assume you still want the nodes you're excluding from color-contrast to run other rules, but if not you could just exclude them completely from axe using the `exclude` option for `axe.run`:

```js
axe.run({
  exclude: ['nodeSelector']
});
```

If you want them in still, it's a bit of a workaround but you might be able configure the color-contrast check to use an `after` function to remove the unwanted nodes.

```js
axe.configure({
  checks: [{
    id: 'color-contrast',
    after: function(results) {
      return results.filter(result => {
        // filter out results you don't want
      });
    }
  }}
});
```

**Comment by nickyblissAviva** [comment]
The after function looks like the best bet meaning we can just remove the few that aren't wanted from that rule but keep the rule for everything else and not exclude the elements completely from the axe tests.
Thank you so much for you help.

**Comment by nickyblissAviva** [comment]
@straker afters results list seems a little odd. The nodes it refers within relatedNodes seem to not be the node that caused the violation so for the page I'm testing with I was getting a 2 different contrast violations one on a node within a form label and one on a form legend but the results are coming back with the form label and the body element.

**Comment by straker** [comment]
Ah, looks like you are correct. `relatedNodes` seems to be the element which has the bad background color and not the node that caused the issue. You still might be able to get away with it if you filter by specific properties of the `data` object (so the node you want to filter has a certain `fgColor`, `bgColor`, `fontSize` that might identify it as unique). 

Other than that, I'm unsure how to accomplish what you need.  Sorry bout that.

**Comment by nickyblissAviva** [comment]
Seems an odd thing to return when the tests and violations deal with the node causing the error itself.
Its not ideal but since we are using Cypress I might have to filter the violations out as they come back into Cypress which isn't ideal as I wanted to get something that didn't rely on Cypress.

**Comment by mohanraj-r** [comment]
@straker I am also looking for the ability to exclude selected rule-components combinations. So the summary here is that it is easier/possible to do it in post-processing after getting the results from axe.run()? Can you please confirm?

**Comment by nickyblissAviva** [comment]
@mohanraj-r I've now got a filtering function setup in Cypress that is taking the results out of axe and pruning out any false positives before sending the results on to be displayed/reported.
Seems to work but isn't ideal as it would have been nice to have been able to run the validation outside of Cypress if required as part of our automated testing setup.
